File: 1239. Maximum Length of a Concatenated String with Unique Characters/maxLength.cpp
class Solution
{
public:
    unordered_map<string, int> dp;
    bool hasDuplicate(string &s1, string &s2)
    {

        vector<int> freq(26, 0);
        for (auto c : s1)
        {
            if (freq[c - 'a'] > 0)
                return true;
            freq[c - 'a']++;
        }
        for (auto c : s2)
        {
            if (freq[c - 'a'] > 0)
                return true;
        }
        return false;
    }
    int solve(int ind, int n, string temp, vector<string> &nums)
    {
        if (ind >= n)
        {
            return temp.length();
        }
        if (dp.find(temp) != dp.end())
            dp[temp];
        int take = 0;
        int skip = 0;
        if (hasDuplicate(nums[ind], temp))
        {
            skip = solve(ind + 1, n, temp, nums);
        }
        else
        {
            skip = solve(ind + 1, n, temp, nums);

            take = solve(ind + 1, n, temp + nums[ind], nums);
        }

        return dp[temp] = max(take, skip);
    }
    int maxLength(vector<string> &arr)
    {

        int n = arr.size();
        string temp = "";
        return solve(0, n, temp, arr);
    }
};

File: 1239. Maximum Length of a Concatenated String with Unique Characters/maxLength.java
class Solution {
    Map<String, Integer> dp;

    public boolean hasDuplicate(String s1, String s2) {
        int[] freq = new int[26];
        for (char c : s1.toCharArray()) {
            if (freq[c - 'a'] > 0) {
                return true;
            }
            freq[c - 'a']++;
        }
        for (char c : s2.toCharArray()) {
            if (freq[c - 'a'] > 0) {
                return true;
            }
        }
        return false;
    }

    public int solve(int ind, int n, String temp, List<String> nums) {
        if (ind >= n) {
            return temp.length();
        }
        if (dp.containsKey(temp)) {
            return dp.get(temp);
        }
        int take = 0;
        int skip = 0;
        if (hasDuplicate(nums.get(ind), temp)) {
            skip = solve(ind + 1, n, temp, nums);
        } else {
            skip = solve(ind + 1, n, temp, nums);
            take = solve(ind + 1, n, temp + nums.get(ind), nums);
        }

        int result = Math.max(take, skip);
        dp.put(temp, result);
        return result;
    }

    public int maxLength(List<String> arr) {
        dp = new HashMap<>();
        String temp = "";
        return solve(0, arr.size(), temp, arr);
    }
}